internal/engine/selectors/selectors.go

```diff
@@ -0,0 +1,372 @@
+// is done only once, we also keep track of the error
+type entityEnvCache struct {
+	once sync.Once
+	env  *cel.Env
+	err  error
+}
+
+// NewEnv creates a new Env struct with the default factories for each entity type. The factories
+// are used on first access to create the CEL environments for each entity type.
+func NewEnv() (*Env, error) {
+	factoryMap := map[minderv1.Entity]celEnvFactory{
+		minderv1.Entity_ENTITY_UNSPECIFIED:  &genericEnvFactory{},
+		minderv1.Entity_ENTITY_REPOSITORIES: &repoEnvFactory{},
+		minderv1.Entity_ENTITY_ARTIFACTS:    &artifactEnvFactory{},
+	}
+
+	entityEnvs := make(map[minderv1.Entity]*entityEnvCache, len(factoryMap))
+	for entity := range factoryMap {
+		entityEnvs[entity] = &entityEnvCache{}
+	}
+
+	return &Env{
+		entityEnvs: entityEnvs,
+		factories:  factoryMap,
+	}, nil
+}
+
+// NewSelectionFromProfile creates a new Selection (compiled CEL programs for that entity type)
+// from a profile
+func (e *Env) NewSelectionFromProfile(
+	entityType minderv1.Entity,
+	profileSelection []*minderv1.Profile_Selector,
+) (Selection, error) {
+	selector := make([]cel.Program, 0, len(profileSelection))
+
+	for _, sel := range profileSelection {
+		ent := minderv1.EntityFromString(sel.GetEntity())
+		if ent != entityType && ent != minderv1.Entity_ENTITY_UNSPECIFIED {
+			continue
+		}
+
+		program, err := e.compileSelectorForEntity(sel.Selector, ent)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compile selector %q: %w", sel.Selector, err)
+		}
+
+		selector = append(selector, program)
+	}
+
+	return &EntitySelection{
+		selector: selector,
+		entity:   entityType,
+	}, nil
+}
+
+// compileSelectorForEntity compiles a selector expression for a given entity type into a CEL program
+func (e *Env) compileSelectorForEntity(selector string, entityType minderv1.Entity) (compiledSelector, error) {
+	env, err := e.envForEntity(entityType)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get environment for entity %v: %w", entityType, err)
+	}
+
+	ast, issues := env.Parse(selector)
+	if issues.Err() != nil {
+		return nil, fmt.Errorf("failed to parse expression %q: %w", selector, issues.Err())
+	}
+
+	checked, issues := env.Check(ast)
+	if issues.Err() != nil {
+		return nil, fmt.Errorf("failed to check expression %q: %w", selector, issues.Err())
+	}
+
+	program, err := env.Program(checked)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create program for expression %q: %w", selector, err)
+	}
+
+	return program, nil
+}
+
+// envForEntity gets the CEL environment for a given entity type. If the environment is not cached,
+// it creates it using the factory for that entity type.
+func (e *Env) envForEntity(entity minderv1.Entity) (*cel.Env, error) {
+	cache, ok := e.entityEnvs[entity]
+	if !ok {
+		return nil, fmt.Errorf("no cache found for entity %v", entity)
+	}
+
+	cache.once.Do(func() {
+		cache.env, cache.err = e.factories[entity].createEnv()
```

Comment:
**nit:** I tend to prefer function types to interfaces when they have a single method, which is usually the case with factories.

(part 3)

```suggestion
		cache.env, cache.err = e.factories[entity]()
```